src/mesh.f90

```diff
@@ -7,43 +7,17 @@ module m_mesh
                       X_EDGE, Y_EDGE, Z_EDGE, &
                       BC_PERIODIC, BC_NEUMANN, BC_DIRICHLET
   use m_field, only: field_t
+  use m_mesh_content
 
   implicit none
 
-  ! Stores geometry information
-  type :: geo_t
-    real(dp), dimension(3) :: d ! size of a cell in each direction (=edge length, distance between centers, distance between vertices)
-    real(dp), dimension(3) :: L ! Global dimensions of the domain in each direction
-  end type
-
-  ! Stores parallel domain related information
-  type :: parallel_t
-    integer :: nrank ! local rank ID
-    integer :: nproc ! total number of ranks/proc participating in the domain decomposition
-    integer, dimension(3) :: nrank_dir ! local rank ID in each direction
-    integer, dimension(3) :: nproc_dir ! total number of proc in each direction
-    integer, dimension(3) :: n_offset  ! number of cells offset in each direction due to domain decomposition
-    integer, dimension(3) :: pnext ! rank ID of the previous rank in each direction
-    integer, dimension(3) :: pprev ! rank ID of the next rank in each direction
-  contains
-    procedure :: is_root ! returns if the current rank is the root rank
-  end type
-
   ! The mesh class stores all the information about the global and local (due to domain decomposition) mesh
   ! It also includes getter functions to access some of its parameters
   type :: mesh_t
-    integer, dimension(3), private :: global_vert_dims ! global number of vertices in each direction without padding (cartesian structure)
-    integer, dimension(3), private :: global_cell_dims ! global number of cells in each direction without padding (cartesian structure)
-
-    integer, dimension(3), private :: vert_dims_padded ! local domain size including padding (cartesian structure)
-    integer, dimension(3), private :: vert_dims ! local number of vertices in each direction without padding (cartesian structure)
-    integer, dimension(3), private :: cell_dims ! local number of cells in each direction without padding (cartesian structure)
-    logical, dimension(3), private :: periodic_BC ! Whether or not a direction has a periodic BC
-    integer, dimension(3, 2), private :: BCs_global
-    integer, dimension(3, 2), private :: BCs
     integer, private :: sz
     type(geo_t), allocatable :: geo ! object containing geometry information
-    type(parallel_t), allocatable :: par ! object containing parallel domain decomposition information
+    class(grid_t), allocatable :: grid ! object containing grid information
```

Comment:
I agree with defining these inside mesh.f90, but that isn't possible due to limitations of the compilers. They flag circular dependencies when doing so (even if there aren't any) just because they compile mesh.f90 at once and can't compile the modules independently. This is the primary reason why all the contents of the mesh object have to be in a separate module (for actual circular dependency) and in a separate file (compiler limitation).